lambdas/src/apis/collections/controllers/collections.ts

```diff
@@ -150,46 +204,117 @@ function buildUrn(protocol: string, contract: string, option: string): string {
 }
 
-async function internalContents(
-  client: SmartContentClient,
-  res: Response,
-  urn: string,
-  selector: (metadata: WearableMetadata) => string | undefined
-): Promise<void> {
-  try {
-    const entity = await fetchEntity(client, urn)
-    if (entity) {
-      const wearableMetadata: WearableMetadata = entity.metadata
-      const hash = findHashForFile(entity, selector(wearableMetadata))
-      if (hash) {
-        const headers: Map<string, string> = await client.pipeContent(hash, res as any as ReadableStream<Uint8Array>)
-        headers.forEach((value: string, key: string) => {
-          res.setHeader(key, value)
-        })
-      } else {
-        res.status(404).send()
-      }
-    }
-  } catch (e) {
-    res.status(500).send(e.message)
-  }
+function validateSize(size: string = DEFAULT_IMAGE_SIZE): size is ValidSize {
+  if (!isValidSize(size)) throw new ServiceError('Invalid size')
+  return true
+}
+
+function getFileHash(entity: Entity, fileName?: string): string {
+  const hash = findHashForFile(entity, fileName)
+  if (!hash) throw new ServiceError(`Hash not found for file ${fileName}`, 404)
+
+  return hash
 }
 
-async function fetchEntity(client: SmartContentClient, urn: string): Promise<Entity | undefined> {
+async function fetchEntity(client: SmartContentClient, urn: string): Promise<Entity> {
   const entities: Entity[] = await client.fetchEntitiesByPointers(EntityType.WEARABLE, [urn])
-  return entities && entities.length > 0 && entities[0].metadata ? entities[0] : undefined
+  if (!(entities && entities.length > 0 && entities[0].metadata)) throw new ServiceError('Entity not found', 404)
+
+  return entities[0]
 }
 
-export function getBodyShapes(representations: WearableMetadataRepresentation[]) {
+function getBodyShapes(representations: WearableMetadataRepresentation[]): string[] {
   const bodyShapes = new Set<WearableBodyShape>()
   for (const representation of representations) {
     for (const bodyShape of representation.bodyShapes) {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
       bodyShapes.add(bodyShape.split(':').pop()!)
     }
   }
   return Array.from(bodyShapes)
 }
 
+function isValidSize(size: string): size is ValidSize {
+  return sizes[size] !== undefined
+}
+
+function getImagePath(root: string, imageRequest: ImageRequest): string {
+  return root + `/images/` + getImageRequestId(imageRequest) + '.png'
+}
+
+// Using this folder structure allow us to find and remove older versions of the same urn (entity)
+function getImageRequestId({ urn, hash, size, rarityBackground }: ImageRequest): string {
+  return `${urn}/${hash}/${rarityBackground ?? 'thumbnail'}-${size}`
+}
+
+// Delete all images that are not the latest version (same hash)
+async function pruneObsoleteImages(root: string, urn: string, hash: string) {
+  const existsFolder = fs.existsSync(root + `/images/` + urn + '/' + hash)
```

Comment:
- are urn valid folder names?
- prevent using sync operations in servers, use promisified versions at all costs when possible to prevent blocking the single thread

Reply:
> are urn valid folder names?
 
 not sure, do you know a breaking case?
 
> prevent using sync operations in servers, use promisified versions at all costs when possible to prevent blocking the single thread

there is no promisifed version for this method (exists) 😞

Reply:
Dang. We are not still there yet.. 

I've found this on stackoverflow but never used it
```ts
function checkFileExists(file) {
  return fs.promises.access(file, fs.constants.F_OK)
           .then(() => true)
           .catch(() => false)
}
```